### DISCUSSION TITLE
How to handle firmware upgrades on remote machines

### DISCUSSION THREAD

**Discussion by drio** [gokrazy/gokrazy#253] [General]
Hello gokrazy users, 

I'd like to know how you handle firmware updates on machines (raspberry pies) for which you don't have physical access to. I know the kernel will be updated (if necessary) when you deploy a new release but I think you need a full OS (RPIOS) to do the firmware update. Is that correct?

Thank you in advance and ... stay krazy.

**Comment by stapelberg** [comment]
Hey @drio 👋

When updating to a new version of gokrazy, all components of the system will be updated.

This includes:
* the Linux kernel (from https://github.com/gokrazy/kernel) 
* the Raspberry Pi firmware (from https://github.com/gokrazy/firmware)
  * i.e. the binary blobs needed for the bootloader.
  * See https://www.raspberrypi.com/documentation/computers/configuration.html#the-boot-folder for more documentation.
* the Raspberry Pi EEPROM (from https://github.com/gokrazy/rpi-eeprom)
  * i.e. the binary blobs stored in the EEPROM.
  * Conceptually, this blob supersedes the `bootcode.bin` file that older models take from the firmware.
  * See https://www.raspberrypi.com/documentation/computers/raspberry-pi.html#raspberry-pi-boot-eeprom for more documentation.

So, no, there is no need to ever run RPIOS on your Pi, you can fully run your Pis with gokrazy :) 

Hope this helps

**Comment by drio** [comment]
Thank you for the explanation Michael. 

IMMO, keeping a full OS up to date (and the firmware) is not an easy task.  Knowing that when you do a `gok update` everything in the pi will be updated to the latest version (only a day behind upstream) makes RPis and gokrazy a strong solution for many scenarios. Not to mention that (if necessary) we can run non-go tools with [docker containers](https://gokrazy.org/packages/docker-containers/) and also have more traditional OS capabilities with [breakglass](https://github.com/gokrazy/breakglass).

I am working on a project that involves deploying devices in different locations to capture wifi metrics to asses the quality of the wifi network. Gokrazy is going to be driving all that. I will report back.

**Reply by damdo** [reply]
Exciting! Yes those are some of the great strenghts of gokrazy! Keep us posted :)
